src/shred_collector/shred_receiver.zig

```diff
@@ -80,76 +80,69 @@ pub const ShredReceiver = struct {
     /// Returns when exit is set to true.
     fn runPacketHandler(
         self: *Self,
-        receivers: []const *Channel(ArrayList(Packet)),
-        response_sender: *Channel(ArrayList(Packet)),
+        receivers: []const *Channel(Packet),
+        response_sender: *Channel(Packet),
         comptime is_repair: bool,
     ) !void {
-        var buf = ArrayList(ArrayList(Packet)).init(self.allocator);
-        while (!self.exit.load(.unordered)) {
+        while (!self.exit.load(.acquire)) {
             var responses = ArrayList(Packet).init(self.allocator);
             for (receivers) |receiver| {
-                try receiver.tryDrainRecycle(&buf);
-                if (buf.items.len > 0) {
-                    const shred_version = self.shred_version.load(.monotonic);
-                    for (buf.items) |batch| {
-                        for (batch.items) |*packet| {
-                            if (is_repair) packet.flags.set(.repair);
-                            try self.handlePacket(packet, &responses, shred_version);
-                        }
-                        try self.unverified_shred_sender.send(batch);
-                    }
+                while (receiver.receive()) |packet| {
+                    var our_packet = packet;
+                    try self.handlePacket(&our_packet, &responses);
+                    if (is_repair) our_packet.flags.set(.repair);
+                    try self.unverified_shred_sender.send(our_packet);
                 } else {
                     std.time.sleep(10 * std.time.ns_per_ms);
                 }
             }
-            if (responses.items.len > 0) {
-                try response_sender.send(responses);
+            for (responses.items) |response| {
+                try response_sender.send(response);
             }
         }
     }
 
     /// Handle a single packet and return
     fn handlePacket(
-        self: *Self,
+        self: Self,
         packet: *Packet,
         responses: *ArrayList(Packet),
-        shred_version: u16,
     ) !void {
         if (packet.size == REPAIR_RESPONSE_SERIALIZED_PING_BYTES) {
-            try self.handlePing(packet, responses);
-            packet.flags.set(.discard);
+            if (try self.handlePing(packet)) |p| try responses.append(p);
         } else {
             const max_slot = std.math.maxInt(Slot); // TODO agave uses BankForks for this
-            if (shouldDiscardShred(packet, self.root_slot, shred_version, max_slot)) {
+            if (shouldDiscardShred(packet, self.root_slot, self.shred_version, max_slot)) {
                 packet.flags.set(.discard);
             }
         }
     }
 
-    /// Handle a ping message and return
-    fn handlePing(self: *Self, packet: *const Packet, responses: *ArrayList(Packet)) !void {
+    /// Handle a ping message and returns the repair message.
+    fn handlePing(self: *const Self, packet: *const Packet) !?Packet {
```

Comment:
Do you expect inlining or RVO to guarantee that the packet will never be copied to the calling context?

Reply:
This will “RVO” not just because of Zig, but because of every real calling convention ever passing non-floats larger than 2 * usize through a return pointer.

As mentioned above, this doing extra copies due to missed elisions is fine, and much better/faster than the alternative of heap allocating.

A more technical argument is that the function will be a bit deeper in the stack than the caller, meaning a new cache set will be used for storing the out buffer. This means no cache pressure is put on the calling function.